feat(sequencer)!: transaction categories on UnsignedTransaction

Lilyjjo opened the pull request:
## Summary
Adds restrictions on what type of `Actions` can be included in the same `UnsignedTransaction`. 

## Background
We want transactions that can affect the validity of other transactions to be ordered last in blocks to reduce the amount of failed transactions we process. These logic changes are the first code changes being made to realize this goal.

## Changes
- Introduced the `Actions` struct to hold valid groupings of `Actions`.
- Changed the `UnsignedTransaction` struct to have private fields.
- Changed the `UnsignedTransaction`'s `action` to be a `Actions` type instead of just a vector of `Actions`.

## Testing
TODO

## Metrics
TODO

## Breaking Changelist
TODO

## Related Issues
Initial steps for #1412

closes #1414, #1416

Lilyjjo commented on crates/astria-composer/src/executor/mod.rs:
```diff
@@ -681,11 +681,12 @@ impl Future for SubmitFut {
+                        .actions(this.bundle.clone().into_actions())
+                        .params(params)
+                        .build()
+                        .expect("failed to build transaction from actions")
```
I don't know if it is acceptable for a malformed transaction to crash this service / if this function call is behind a protected endpoint.

Lilyjjo replied:
The code in the `bundle_factory` is restricted to just work for `SequenceActions` so this shouldn't ever fail, might be better to move the construction of the `UnsignedTransaction` to the bundle code or to have the `bundle_factory` code to keep everything as explicit `SequenceActions`

Lilyjjo replied:
Went the route of putting it into a function

Lilyjjo commented on crates/astria-composer/src/executor/mod.rs:
```diff
@@ -759,11 +760,12 @@ impl Future for SubmitFut {
+                            .actions(this.bundle.clone().into_actions())
+                            .params(params)
+                            .build()
+                            .expect("failed to build transaction from actions")
```
Same here.

SuperFluffy commented on crates/astria-sequencer/src/app/mod.rs:
```diff
@@ -1024,10 +1025,10 @@ impl App {
-                .any(|action| matches!(action, Action::FeeAssetChange(_) | Action::FeeChange(_)));
+            || matches!(signed_tx.group(), Some(ActionGroup::BundlableSudo(_)))
+                && signed_tx.actions().iter().any(|action| {
+                    matches!(action, Action::FeeAssetChange(_) | Action::FeeChange(_))
```
It's idiomatic to implement these as inherent methods on an enum so that this would read like:

```rust
// inherent method
impl Action {
    fn is_fee_asset_change(&self) -> bool {
        matches!(action, Self::FeeAssetChange(_))
    }
    fn is_fee_change(&self) -> bool {
        matches!(action, Self::FeeChange(_))
    }
}

// the function
&& signed_tx.actions().iter().any(|act| act.is_fee_asset_change() || act.is_fee_change())
```

SuperFluffy commented on crates/astria-cli/src/commands/bridge/submit.rs:
```diff
@@ -129,14 +129,17 @@ async fn submit_transaction(
+                .build(),
+        )
+        .build()
+        .expect("failed to build transaction from actions")
```
You are in an environment that exits early on errors. I would return an error here to potentially get some error feedbackk.

SuperFluffy commented on crates/astria-cli/src/commands/sequencer.rs:
```diff
@@ -475,14 +475,17 @@ async fn submit_transaction(
+                .build(),
+        )
+        .build()
+        .expect("failed to build transaction from actions")
```
Same here - you can just return an error, so `wrap_err("...")?` is not more work

SuperFluffy commented on crates/astria-sequencer/src/proposal/commitment.rs:
```diff
@@ -121,13 +121,17 @@ mod test {
-            actions: vec![sequence_action.clone().into(), transfer_action.into()],
-        };
+
+        let tx = UnsignedTransaction::builder()
```
can you make a utility for this? It looks like there is a lot of repetition for constructing these transactions here.

SuperFluffy reviewed (commented):
Looks like an elegant solution to me, and I am always a fan of locking down types.

I think this is ready to be merged, but I would like to see the errors for the wire-type -> domain-type conversion to contain more information.

Fraser999 reviewed (commented):
Initial (incomplete) review.

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_groups.rs:
```diff
@@ -0,0 +1,228 @@
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BundlableGeneral;
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct General;
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BundlableSudo;
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Sudo;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ActionGroup {
+    BundlableGeneral(BundlableGeneral),
+    General(General),
+    BundlableSudo(BundlableSudo),
+    Sudo(Sudo),
+}
+
+impl From<BundlableGeneral> for ActionGroup {
+    fn from(val: BundlableGeneral) -> ActionGroup {
+        ActionGroup::BundlableGeneral(val)
+    }
+}
+
+impl From<General> for ActionGroup {
+    fn from(val: General) -> ActionGroup {
+        ActionGroup::General(val)
+    }
+}
+
+impl From<BundlableSudo> for ActionGroup {
+    fn from(val: BundlableSudo) -> ActionGroup {
+        ActionGroup::BundlableSudo(val)
+    }
+}
+
+impl From<Sudo> for ActionGroup {
+    fn from(val: Sudo) -> ActionGroup {
+        ActionGroup::Sudo(val)
+    }
+}
```
Note that this was my original suggestion in order to document the group membership for each action.

Doing it via the `Action` enum moves this property away from each action.

Fraser999 replied:
Well, the associated consts are invisible in rustdoc since they're private, but that's easily remedied I guess.  Likewise, the `BelongsToGroup` trait is private, so it's also invisible.

If we make the `BelongsToGroup` trait `pub` but remove the associated consts, the docs are still pretty unhelpful since each action will have the same generated docs, like:
```
impl BelongsToGroup for BridgeLockAction
fn belongs_to_group(&self) -> ActionGroup
```

We could maybe change the trait to:
```rust
pub trait BelongsToGroup: Sealed {
    const GROUP: ActionGroup;
}
```
so that generated docs look like:
```
impl BelongsToGroup for BridgeLockAction
const GROUP: ActionGroup = ActionGroup::BundlableGeneral
```
It would mean the trait can't be implemented on the `Action` enum, but I think that's a good thing if the whole reason for the trait is to support rustdoc documentation showing which group the given action belongs to, and that is impossible for an enum which can be one of several groups.

SuperFluffy replied:
Yeah, I like that.

Lilyjjo commented:
Since last set of reviews:
- Rewrote action_group.rs trait setup in suggested way
- Changed `UnsignedTransaction()` from using a `Builder` to use a `new()` constructor
- Made error handling more verbose

noot commented on crates/astria-composer/src/executor/bundle_factory/mod.rs:
```diff
@@ -74,6 +76,12 @@ impl SizedBundle {
         }
     }
 
+    #[allow(clippy::panic)] // method is expected to never panic as an invariant of the type
```
because the actions are only ever of the `SequenceAction` type?

Lilyjjo replied:
made more descriptive

noot commented on crates/astria-core/src/protocol/test_utils.rs:
```diff
@@ -104,16 +104,18 @@ impl ConfigureSequencerBlock {
-            };
+            )
+            .expect(
+                "failed to build unsigned transaction, should be okay since only sequence actions",
```
```suggestion
                "should be able to build unsigned transaction since only sequence actions are contained",
```

noot commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+        }
+    }
+
+    pub(super) fn from_list_of_actions(actions: Vec<Action>) -> Result<Self, Error> {
```
```suggestion
    pub(super) fn try_from_list_of_actions(actions: Vec<Action>) -> Result<Self, Error> {
```

since it returns a result

noot commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ActionGroup {
+    BundlableGeneral(BundlableGeneral),
+    General(General),
```
```suggestion
    UnbundleableGeneral(UnbundleableGeneral),
```

I think `General` is a bit unclear, would prefer to explicitly say unbundleable (and maybe even rename BundleableGeneral to just General?)

Lilyjjo replied:
I agree, I'll change the names to note the unbundleable instead!

noot commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+    BundlableGeneral(BundlableGeneral),
+    General(General),
+    BundlableSudo(BundlableSudo),
+    Sudo(Sudo),
```
```suggestion
    UnbundleableSudo(UnbundleableSudo),
```

same here

noot commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct BundlableGeneral;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct General;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct BundlableSudo;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Sudo;
```
are these ever used anywhere? they seem unnecessary, can the `ActionGroup` enum variant be used?

Lilyjjo replied:
You're right, this was left over from a different implementation route. Good catch!

noot commented on crates/astria-sequencer/src/app/test_utils.rs:
```diff
@@ -223,20 +223,21 @@ pub(crate) fn mock_tx(
+            .chain_id("test")
+            .build(),
+    )
+    .expect("Failed to build unsigned transaction");
```
```suggestion
    .expect("can build unsigned transaction");
```

SuperFluffy replied:
The semantics of @noot suggestion are better, but the new statement is somewhat vacuous. But since you want to create `mock_tx` it's clear from the context that you probably expect this to work as inputs to other tests:
```suggestion
    .unwrap();
```

noot commented on crates/astria-sequencer/src/app/tests_app/mempool.rs:
```diff
@@ -48,23 +48,23 @@ async fn trigger_cleaning() {
+            .chain_id("test")
+            .build(),
+    )
+    .expect("failed to build unsigned transaction")
```
```suggestion
    .expect("can build unsigned transaction")
```

noot replied:
same for other in this file

SuperFluffy replied:
As above: this does not provide info as to why you expect this to have worked. If you think it's trivially clear why this should have worked, prefer `unwrap()`.

SuperFluffy commented on crates/astria-sequencer/src/app/mod.rs:
```diff
@@ -1024,10 +1025,11 @@ impl App {
-                .actions()
-                .iter()
-                .any(|action| matches!(action, Action::FeeAssetChange(_) | Action::FeeChange(_)));
+            || matches!(signed_tx.group(), Some(ActionGroup::BundlableSudo(_)))
```
It's convetion to create inherent methods for checks like this (similar to the check below; I think I commented this before?):

```rust
impl SignedTransaction {
    // NOTE: I used `contains_*` rather than `is_` because it read weird; maybe you can come up with better naming?
    fn contains_bundleable_sudo_actions(&self) -> bool {
        matches!(self.group(), ActionGroup::BundleSudo(_))
    }
}
```

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
```
Note on all these types:

Do we really need to expose them publicly? Looking at the sequencer code, this is only needed in the `matches!` statement in mempool?

Instead of exposing all these groups publicly, can we just provide inherents methods like `UnsignedTransaction::contains_bundleable_sudo_actions` (or potentially better names)?

This would significantly reduce the public surface area this patch adds.

Lilyjjo replied:
We might need access to them in a future PR for sorting transactions in the block building process, I can make them private for now

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+    Sudo(Sudo),
+}
+
+impl fmt::Display for ActionGroup {
```
These display impls are essentially their `Debug` counterparts. Can you make them more human readable?

Just something like `bundleable general` instead of snakecase.

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+}
+
+#[derive(Debug, thiserror::Error)]
+enum ErrorKind {
```
Noted further down on the custom `Display` implementation on public `Error` type: lean on the `error` attribute of the `thiserror` proc macro to generate that code for you. I.e. use struct variants, for example like so:

```rust
#[derive(Debug, thiserrror::Error)]
enum ErrorKind {
    #[error("actions of different types are not permitted; input contained `{group_one}` and `{group_two}`")]
    Mixed {
        group_one: ActionGroup,
        group_two: ActionGroup,
    }
}
```

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+    NotBundleable,
+}
+
+#[derive(Debug)]
```
For consistency with the rest of the crate:
```rust
#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub struct Error(ErrorKind);
```
And put the context into the kinds.

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+        };
+
+        // assert size constraints on non-bundlable action groups
+        if matches!(group, ActionGroup::General(_) | ActionGroup::Sudo(_)) && actions.len() > 1 {
```
Create an inherent method for this: `ActionGroup::is_bundleable`

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs:
```diff
@@ -0,0 +1,491 @@
+    }
+}
+
+#[cfg(test)]
```
Let's move these into a `tests.rs` crate, i.e.:

```
action_group/
  mod.rs
  tests.rs
```

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs:
```diff
@@ -198,7 +203,12 @@ impl SignedTransaction {
+    }
+
+    #[must_use]
+    pub fn group(&self) -> Option<ActionGroup> {
```
Noted elsewhere: I think we can avoid exposing this publicly for the time being by providing `contains_xxx_group_actions` or `is_xxx_group_actions`

SuperFluffy commented on crates/astria-sequencer/src/app/tests_app/mempool.rs:
```diff
@@ -147,24 +147,23 @@ async fn do_not_trigger_cleaning() {
+            .chain_id("test")
+            .build(),
+    )
+    .expect("failed to build unsigned transaction")
```
As above: this does not provide info as to why you expect this to have worked. If you think it's trivially clear why this should have worked, prefer `unwrap()`.

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action.rs:
```diff
@@ -252,6 +260,30 @@ impl TryFrom<raw::Action> for Action {
     }
 }
 
+pub(super) trait ActionName {
```
Please remove this. All of these should be done via the `Protobuf` trait implemented at the root of the crate - I intend to rename it to `DomainType`.

SuperFluffy replied:
Discussed offline: this should flow through `Protobuf::full_name` but is out of scope for this PR. @Lilyjjo will create a PR to follow up on this.

Lilyjjo replied:
Tracked in a note in the code and here: #1567

Lilyjjo commented on crates/astria-sequencer/src/app/snapshots/astria_sequencer__app__tests_breaking_changes__app_execute_transaction_with_every_action_snapshot.snap:
```diff
@@ -1,39 +1,38 @@
-  173,
-  132,
-  77,
+   96,
```
```suggestion
  96,
```

Fraser999 commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group/tests.rs:
```diff
@@ -0,0 +1,244 @@
+        }),
+    ];
+
+    assert_eq!(
+        Actions::try_from_list_of_actions(actions)
+            .unwrap_err()
+            .to_string(),
+        "attempted to create bundle with non bundleable `ActionGroup` type: unbundleable sudo"
+    );
```
```suggestion
    let error_kind = Actions::try_from_list_of_actions(actions).unwrap_err().0;
    assert!(
        matches!(error_kind, ErrorKind::NotBundleable { .. }),
        "expected ErrorKind::NotBundleable, got {error_kind:?}"
    );
```
Similar suggestion for following two tests also.

Fraser999 commented on crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs:
```diff
@@ -381,35 +417,50 @@ enum UnsignedTransactionErrorKind {
-    fn new() -> Self {
+impl UnsignedTransactionBuilder {
+    #[must_use]
+    #[allow(clippy::new_without_default)]
```
We should be able to just derive `Default` on the builder rather than doing this `allow` I think.

Fraser999 commented on crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs:
```diff
@@ -381,35 +417,50 @@ enum UnsignedTransactionErrorKind {
 }
 
-pub struct TransactionParamsBuilder<TChainId = std::borrow::Cow<'static, str>> {
+pub struct UnsignedTransactionBuilder<TChainId = std::borrow::Cow<'static, str>> {
```
Having a generic parameter seems a little excessive to me.  I think it'd be fine to just have `chain_id: String` and have the setter as `pub fn chain_id<T: Into<String>>(self, chain_id: T)`

Fraser999 commented on crates/astria-sequencer/src/app/tests_breaking_changes.rs:
```diff
@@ -307,3 +351,175 @@ async fn app_execute_transaction_with_every_action_snapshot() {
+            .chain_id("test")
+            .try_build()
+            .unwrap_err()
+            .to_string(),
+        "attempted to create bundle with non bundleable `ActionGroup` type: unbundleable sudo"
```
As before, I think these tests should match on the `ErrorKind` variant rather than on the `Display` output of the error.

Lilyjjo replied:
I made the whole `ActionGroup` enum public in the process to achieve this.

Fraser999 reviewed (commented):
Other than the aspect discussed offline, just nitpicks from me.

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group/mod.rs:
```diff
@@ -0,0 +1,201 @@
+
+impl Error {
+    #[must_use]
+    pub fn kind(&self) -> &ErrorKind {
```
private?

Lilyjjo replied:
I didn't know how to access the error for a cleaner testing style that Fraser recommended without making public: https://github.com/astriaorg/astria/blob/3e79a1c7794fe1fd78dcc25ae72de8b8f8b399c7/crates/astria-sequencer/src/app/tests_breaking_changes.rs#L525

SuperFluffy replied:
I think these tests are misplaced.

If I see it correctly, you want to check if the bundling logic works correctly, right? The correct place to test this is where the bundling is implemented, i.e. here in `astria-core`. `astria-sequencer` as a downstream consumer should be able to rely on the implementation being correct.

From a quick glance I could not see if there is any logic that branches on the error kind and subsequently takes a different path.

SuperFluffy replied:
There really isn't, but that's why we have a review process and rely on authors to note breaking changes.

But it is implicitly understood that changing behavior of fundamental types in `astria-core` cause immediate breakage in `astria-sequencer`.

Copying tests and publicly exposing implementation deatils for what's effectively a unit test in other parts of the code base just adds extra churn without giving you extra safety.

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group/mod.rs:
```diff
@@ -0,0 +1,201 @@
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ErrorKind {
```
private? Do you need to check why bundling failed and take action based on it?

Lilyjjo replied:
Responded on the comment above

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs:
```diff
@@ -381,59 +417,69 @@ enum UnsignedTransactionErrorKind {
         raw::UnsignedTransaction::type_url()
     )]
     DecodeAny(#[source] prost::DecodeError),
+    #[error("`actions` field does not make a valid `ActionGroup`")]
```
`ActionGroup` is not a concept outside of the crate:
```suggestion
    #[error("`actions` field does not form a valid group of actions")]
```

SuperFluffy commented on crates/astria-sequencer/src/app/tests_breaking_changes.rs:
```diff
@@ -307,3 +352,178 @@ async fn app_execute_transaction_with_every_action_snapshot() {
+// be marked as breaking.
+#[allow(clippy::too_many_lines)]
+#[tokio::test]
+async fn app_transaction_bundle_categories() {
```
I commented elsewhere, the tests here test implementation details of `astria-core` and are redundant.

It is implicitly understood that changes in behavior of `astria-core` are immediate breaking for `astria-sequencer` because they are a stepping stone.

Lilyjjo replied:
Ok, will remove.

SuperFluffy commented on crates/astria-composer/src/executor/bundle_factory/mod.rs:
```diff
@@ -112,6 +130,7 @@ impl SizedBundle {
     }
 
     /// Consume self and return the underlying buffer of actions.
+    #[cfg(test)]
```
It's inadvisable to test code paths that only exists for tests - you are likely to check wrong behavior.

I think you can remove `into_actions` and, in the tests, just use `to_actions` instead.

SuperFluffy commented on crates/astria-composer/src/executor/bundle_factory/tests.rs:
```diff
@@ -333,4 +333,26 @@ mod bundle_factory {
     }
+
+    #[test]
+    fn unsigned_transaction_construction() {
```
```suggestion
    fn transaction_construction_does_not_panic() {
```

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action.rs:
```diff
@@ -169,6 +169,14 @@ impl Action {
         Some(transfer_action)
     }
+
+    pub fn is_fee_asset_change(&self) -> bool {
```
Let's make a tracking issue to add unit tests for these.

Lilyjjo replied:
created and commented in code: https://github.com/astriaorg/astria/issues/1593

SuperFluffy commented on crates/astria-composer/src/executor/bundle_factory/mod.rs:
```diff
@@ -74,6 +75,23 @@ impl SizedBundle {
+    /// Constructs an [`UnsignedTransaction`] from the actions contained in the bundle and `params`.
+    // Method is expected to never panic because only `SequenceActions` are added to the bundle,
+    // which should produce a valid variant of the `ActionGroup` type.
+    #[allow(clippy::panic)]
```
Heads-up: with https://github.com/astriaorg/astria/pull/1561 being merged we switch everything to `expect` and require a `reason`:
```suggestion
#[expect(clippy::panic, reason = "method is expected to never panic because only `SequenceActions` are added to the bundle, which should produce a valid variant of the `ActionGroup` type; this is checked by `tests::transaction_construction_should_not_panic")]
```

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group/mod.rs:
```diff
@@ -0,0 +1,196 @@
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(super) enum ActionGroup {
+    General,
+    UnbundleableGeneral,
+    Sudo,
+    UnbundleableSudo,
```
Bikeshed:

@noot left some comments on these where `Bundleable*` was removed and `Unbundleable*` introduced.

My preference would have been to go all out and use `General -> BundleableGeneral` and `Sudo -> BundleableSudo` (and keep `Unbundleable*`) so that this is always spelled out explicitly.

But I am going to leave the final call on this one to you.

Lilyjjo replied:
Changed to BundleableXXX. Agree it's easier to understand

SuperFluffy commented on crates/astria-core/src/protocol/transaction/v1alpha1/action_group/mod.rs:
```diff
@@ -0,0 +1,196 @@
+impl fmt::Display for ActionGroup {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ActionGroup::General => write!(f, "general"),
```
Just leaving a note that should you go with my bikeshedding to rename prefix with `Bundleable*`, to also change the display name.

SuperFluffy reviewed (approved):
Thank you for going going through the many review comments, online and offline!
